logstreams/src/main/java/io/zeebe/logstreams/state/ReplicationController.java

```diff
@@ -0,0 +1,246 @@
+
+      int len;
+      while ((len = checkSumInputStream.read(readBuffer)) != -1) {
+        byteArrayStream.write(readBuffer, 0, len);
+      }
+
+      final byte[] content = byteArrayStream.toByteArray();
+      final long checksum = checkSumInputStream.getChecksum().getValue();
+
+      replication.replicate(
+          new SnapshotChunkImpl(
+              snapshotPosition, totalCount, snapshotChunkFile.getName(), checksum, content));
+    } catch (IOException ioe) {
+      LOG.error(
+          "Unexpected error on reading snapshot chunk from file '{}'.", snapshotChunkFile, ioe);
+    }
+  }
+
+  /** Registering for consuming snapshot chunks. */
+  void consumeReplicatedSnapshots() {
+    replication.consume((this::consumeSnapshotChunk));
+  }
+
+  /**
+   * This is called by the snapshot replication implementation on each snapshot chunk
+   *
+   * @param snapshotChunk the chunk to consume
+   */
+  private void consumeSnapshotChunk(SnapshotChunk snapshotChunk) {
+    final long snapshotPosition = snapshotChunk.getSnapshotPosition();
+    final String snapshotName = Long.toString(snapshotPosition);
+    final String chunkName = snapshotChunk.getChunkName();
+
+    if (storage.existSnapshot(snapshotPosition)) {
+      LOG.debug("Ignore snapshot chunk {}, snapshot {} already exist.", chunkName, snapshotName);
+      return;
+    }
+
+    final long snapshotCounter =
+        receivedSnapshots.computeIfAbsent(snapshotPosition, k -> START_VALUE);
+    if (snapshotCounter == INVALID_SNAPSHOT) {
+      LOG.debug(
+          "Ignore snapshot chunk {}, because snapshot {} is marked as invalid.",
+          chunkName,
+          snapshotName);
+      return;
+    }
+
+    final long expectedChecksum = snapshotChunk.getChecksum();
+    final long actualChecksum = createActualChecksum(snapshotChunk);
+
+    if (expectedChecksum != actualChecksum) {
+      markSnapshotAsInvalid(snapshotChunk);
+      LOG.warn(
+          "Expected to have checksum {} for snapshot chunk file {} ({}), but calculated {}",
+          expectedChecksum,
+          chunkName,
+          snapshotName,
+          actualChecksum);
+      return;
+    }
+
+    final File tmpSnapshotDirectory = storage.getTmpSnapshotDirectoryFor(snapshotName);
+    if (!tmpSnapshotDirectory.exists()) {
+      tmpSnapshotDirectory.mkdirs();
+    }
+
+    final File snapshotFile = new File(tmpSnapshotDirectory, chunkName);
+    if (snapshotFile.exists()) {
+      LOG.debug("Received a snapshot file which already exist '{}'.", snapshotFile);
+      return;
+    }
+
+    LOG.debug("Consume snapshot chunk {}", chunkName);
+    writeReceivedSnapshotChunk(snapshotChunk, tmpSnapshotDirectory, snapshotFile);
+  }
+
+  private long createActualChecksum(SnapshotChunk snapshotChunk) {
+    final byte[] content = snapshotChunk.getContent();
+    final CRC32 crc32 = new CRC32();
+    crc32.update(content);
+    return crc32.getValue();
```

Comment:
I would prefer to use the same method to create the checksums at the sender and the receiver.